doc/neps/nep-0041-improved-dtype-support.rst

```diff
@@ -0,0 +1,706 @@
+would percolate from the datatype to the ndarray.
+
+The interaction with other scalars would likely be defined through::
+
+    >>> np.common_type(np.float64, Unit)
+    Unit[np.float64](dimensionless)
+
+Ufunc output datatype determination can be more involved than for simple
+numerical dtypes since there is no "universal" output type::
+
+    >>> np.multiply(meters, seconds).dtype != np.result_type(meters, seconds)
+
+In fact ``np.result_type(meters, seconds)`` must error without context
+of the operation being done.
+This example highlights how the specific ufunc loop
+(loop with known, specific DTypes as inputs), has to be able to to make
+certain decisions before the actual calculation can start.
+
+
+
+Implementation
+--------------
+
+Plan to Approach the Full Refactor
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To address these issues in NumPy and enable new datatypes,
+multiple development stages are required:
+
+* Phase I: Restructure and extend the datatype infrastructure (This NEP)
+
+  * Organize Datatypes like normal Python classes [`PR 15508`]_
+
+* Phase II: Incrementally define or rework API
+
+  * Create a new and easily extensible API for defining new datatypes
+    and related functionality. (NEP 42)
+
+  * Incrementally define all necessary functionality through the new API (NEP 42):
+
+    * Defining operations such as ``np.common_type``.
+    * Allowing to define casting between datatypes.
+    * Add functionality necessary to create a numpy array from Python scalars
+      (i.e. ``np.array(...)``).
+    * …
+
+  * Restructure how universal functions functions to (NEP 43):
+
+    * make it possible to allow a `~numpy.ufunc` such as ``np.add`` to be
+      extended by user defined datatypes such as Units.
+
+    * allow efficient lookup for the correct implementation for user defined
+      datatypes.
+
+    * enable reuse of existing code. Units should be able to use the
+      normal math loops and add additional logic to determine output type.
+
+* Phase III: Growth of NumPy and Scientific Python Ecosystem capabilities:
+
+  * Cleanup of legacy behaviour where it is considered buggy or undesirable.
+  * Provide a path to define new datatypes from Python.
+  * Assist the community in creating types such as Units or Categoricals
+  * Allow strings to be used in functions such as ``np.equal`` or ``np.add``.
+  * Remove legacy code paths within NumPy to improve long term maintainability
+
+This document serves as a basis for phase I and provides the vision and
+motivation for the full project.
+While of limited use to the end user, the first step involves a conceptional
+cleanup of the current datatype system. It provides a more "pythonic" datatype
+Python type object, with a clear class hierarchy.
+
+The second phase is the incremental creation of all APIs necessary to define
+fully featured datatypes and reorganization of the NumPy datatype system.
+Phase I concerns mainly private APIs, while phase II is expected to include
```

Comment:
The wording here might invite questions about this development plan - might want to be less specific and just mention that phase 2 is concerned with the design and development of the public API.